### PR TITLE
Create temporary files in book folder if needed (BL-3954)

### PR DESCRIPTION
**Pull request by StephenMcConnel**
This preserves the proper permissions for access over a network
file share.

This is an alternative to https://github.com/BloomBooks/BloomDesktop/pull/1322.

---
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/1327)

### CODE REVIEW

File: src/BloomExe/Book/Book.cs [hunk 1]
```diff
@@ -2062,13 +2062,18 @@ private void WriteLanguageDisplayStyleSheet( )
 
 			using (var temp = TempFile.WithExtension(".css"))
 			{
-				RobustFile.WriteAllText(temp.Path,
+				// Use a temporary file pathname in the current book's folder.  This is needed to ensure proper permissions are granted
+				// to the resulting file later after FileUtils.ReplaceFileWithUserInteractionIfNeeded is called.  That method may call
+				// File.Replace which replaces both the file content and the file metadata (permissions).  The result of that if we use
+				// the user's temp directory is described in http://issues.bloomlibrary.org/youtrack/issue/BL-3954.
+				var tempPath = Path.Combine(_storage.FolderPath, Path.GetFileName(temp.Path));
```
Comment (hatton) [inline src/BloomExe/Book/Book.cs]: Isn't temp.Path more than just a file name? So what does tempPath end up being? Something like c:\....docs\bloom\mycollection\mybook\c:\user\joe\local\temp? If so, how about 

`tempPath = Path.Combine( ...., Guid.NewGuid().ToString())`

File: src/BloomExe/Book/Book.cs [hunk 2]
```diff
@@ -2062,13 +2062,18 @@ private void WriteLanguageDisplayStyleSheet( )
 
 			using (var temp = TempFile.WithExtension(".css"))
 			{
-				RobustFile.WriteAllText(temp.Path,
+				// Use a temporary file pathname in the current book's folder.  This is needed to ensure proper permissions are granted
```
Comment (hatton) [inline src/BloomExe/Book/Book.cs]: I'm wondering if there is a natural way to capture this pattern you're doing... like

``` using(var temp = TempFile.InFolder(_storage.FolderPath).WithExtension(".css"); 

Though in this case at least, there isn't a need for the extension if I'm not mistaken, as we later copy it over.
If that is always going to be the case (and maybe it is), then just implementing TempFile.InFolder() would be enough.

**Comment by StephenMcConnel** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 2069 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327#-KWtgnGXf5tRABs21w06-r1-2069:-KWtsgyucjiaHsmPOMLc:b-40a1ex) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/58c8639d025817a88da844894798770257fd3bb8/src/BloomExe/Book/Book.cs#L2069)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Isn't temp.Path more than just a file name? So what does tempPath end up being? Something like c:\....docs\bloom\mycollection\mybook\c:\user\joe\local\temp? If so, how about 
> 
> 
> 
> `tempPath = Path.Combine( ...., Guid.NewGuid().ToString())`
</details>
The filename portion is something like tmpAB123F.tmp.  Not quite as wild as a guid, but more complicated than a simple name.

---

*[src/BloomExe/Book/BookStorage.cs, line 245 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327#-KWtgnGXf5tRABs21w07-r1-245:-KWtt0SEzhhSOjty6kDO:b333gf1) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/58c8639d025817a88da844894798770257fd3bb8/src/BloomExe/Book/BookStorage.cs#L245)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> How is this related?
</details>
To prevent the temporary file from hanging around since nothing removes it otherwise.  (A File.Copy is used a few lines above.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327)*

**Comment by StephenMcConnel** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 2065 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327#-KWtgnGXf5tRABs21w06-r1-2065:-KWttVuXWkQ22yxMORGM:bi1nxex) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/58c8639d025817a88da844894798770257fd3bb8/src/BloomExe/Book/Book.cs#L2065)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> I'm wondering if there is a natural way to capture this pattern you're doing... like
> 
> 
> 
> ``` using(var temp = TempFile.InFolder(_storage.FolderPath).WithExtension(".css"); 
> 
> 
> 
> Though in this case at least, there isn't a need for the extension if I'm not mistaken, as we later copy it over.
> 
> If that is always going to be the case (and maybe it is), then just implementing TempFile.InFolder() would be enough.
</details>
That may well be worth doing.  It requires a libpalaso PR before this PR can be finalized.  Since it's late Friday as I write this, we can decide on Monday what to do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327)*

**Comment by StephenMcConnel** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 2065 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327#-KWtgnGXf5tRABs21w06-r1-2065:-KX7AeyIC9GzmjbbPxz7:b-83e3dq) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/58c8639d025817a88da844894798770257fd3bb8/src/BloomExe/Book/Book.cs#L2065)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary>
> That may well be worth doing.  It requires a libpalaso PR before this PR can be finalized.  Since it's late Friday as I write this, we can decide on Monday what to do.
</details>
I've added a method TempFile.InFolderOf() to SIL.Core, which is currently available for review.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327)*

**Comment by StephenMcConnel** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 2065 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327#-KWtgnGXf5tRABs21w06-r1-2065:-KXCB6kpFmmZwlcGz7vB:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/58c8639d025817a88da844894798770257fd3bb8/src/BloomExe/Book/Book.cs#L2065)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary>
> I've added a method TempFile.InFolderOf() to SIL.Core, which is currently available for review.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327)*

File: src/BloomExe/Book/BookStorage.cs [hunk 3]
```diff
@@ -335,13 +337,26 @@ private void AssertIsAlreadyInitialized()
...
 			return XmlHtmlConverter.SaveDOMAsHtml5(dom.RawDom, tempPath);
 		}
 
+		/// <summary>
+		/// Get a temporary file pathname in the current book's folder.  This is needed to ensure proper permissions are granted
+		/// to the resulting file later after FileUtils.ReplaceFileWithUserInteractionIfNeeded is called.  That method may call
+		/// File.Replace which replaces both the file content and the file metadata (permissions).  The result of that if we use
+		/// the user's temp directory is described in http://issues.bloomlibrary.org/youtrack/issue/BL-3954.
+		/// </summary>
+		private string GetTempHtmlFileName()
```
Comment (hatton) [inline src/BloomExe/Book/BookStorage.cs]: How about GetNameForATempFileInStorageFolder()  . There's no .html to this.

**Comment by StephenMcConnel** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/BloomExe/Book/BookStorage.cs, line 353 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327#-KWtgnGXf5tRABs21w07-r2-353:-KXCGCsjATAIzfNmKo8T:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/3c202c9d2e8e6f1801370dd3177f10ae966608ba/src/BloomExe/Book/BookStorage.cs#L353)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> How about GetNameForATempFileInStorageFolder()  . There's no .html to this.
</details>
Done.

---

*[src/BloomExe/ImageProcessing/ImageUtils.cs, line 291 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327#-KWtgnGXf5tRABs21w08-r2-291:-KXCGPhYaRIFK7IFCg0-:b-apqjnz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/3c202c9d2e8e6f1801370dd3177f10ae966608ba/src/BloomExe/ImageProcessing/ImageUtils.cs#L291)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> This was a pre-existing problem, but this should be "temp", not "tempPath". It's not a path.
</details>
Done.  I actually had this change in my original PR, but went back to minimizing the difference from the original code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1327)*

**Pull request merged**